Description: Eliminate Rust heap allocations and expose Table offsets.
    This eliminates latency spikes caused by heap allocations during
    real-time serialization, and enables custom downstream reflection
    tools to directly inspect Table positions and raw byte slices.
    Implemented by declaring arrayvec as a library dependency in
    rust/flatbuffers/Cargo.toml, replacing heap-allocated Vec types
    inside FlatBufferBuilder (field_locs, written_vtable_revpos,
    strings_pool) with fixed-size ArrayVec structures in
    rust/flatbuffers/src/builder.rs, and exposing the buf and loc
    fields as pub in Table.


diff --git a/rust/flatbuffers/Cargo.toml b/rust/flatbuffers/Cargo.toml
index 9dba11a..b60cba9 100644
--- a/rust/flatbuffers/Cargo.toml
+++ b/rust/flatbuffers/Cargo.toml
@@ -17,7 +17,8 @@ std = []
 serialize = ["serde"]
 
 [dependencies]
-bitflags = "2.8.0"
+bitflags = "1.2.1"
+arrayvec = "0.7.4"
 serde = { version = "1.0", optional = true }
 
 [build-dependencies]
diff --git a/rust/flatbuffers/src/builder.rs b/rust/flatbuffers/src/builder.rs
index d8a6e81..466762a 100644
--- a/rust/flatbuffers/src/builder.rs
+++ b/rust/flatbuffers/src/builder.rs
@@ -16,6 +16,7 @@
 
 #[cfg(not(feature = "std"))]
 use alloc::{vec, vec::Vec};
+use arrayvec::ArrayVec;
 use core::cmp::max;
 use core::convert::Infallible;
 use core::fmt::{Debug, Display};
@@ -131,15 +132,16 @@ pub struct FlatBufferBuilder<'fbb, A: Allocator = DefaultAllocator> {
     allocator: A,
     head: ReverseIndex,
 
-    field_locs: Vec<FieldLoc>,
-    written_vtable_revpos: Vec<UOffsetT>,
+    // TODO(Adam): Make vectors generic.
+    field_locs: ArrayVec<FieldLoc, 100>,
+    written_vtable_revpos: ArrayVec<UOffsetT, 20>,
 
     nested: bool,
     finished: bool,
 
     min_align: usize,
     force_defaults: bool,
-    strings_pool: Vec<WIPOffset<&'fbb str>>,
+    strings_pool: ArrayVec<WIPOffset<&'fbb str>, 20>,
 
     _phantom: PhantomData<&'fbb ()>,
 }
@@ -189,15 +191,15 @@ impl<'fbb, A: Allocator> FlatBufferBuilder<'fbb, A> {
             allocator,
             head,
 
-            field_locs: Vec::new(),
-            written_vtable_revpos: Vec::new(),
+            field_locs: ArrayVec::new(),
+            written_vtable_revpos: ArrayVec::new(),
 
             nested: false,
             finished: false,
 
             min_align: 0,
             force_defaults: false,
-            strings_pool: Vec::new(),
+            strings_pool: ArrayVec::new(),
 
             _phantom: PhantomData,
         }
diff --git a/rust/flatbuffers/src/table.rs b/rust/flatbuffers/src/table.rs
index d3c296b..1847e53 100644
--- a/rust/flatbuffers/src/table.rs
+++ b/rust/flatbuffers/src/table.rs
@@ -20,8 +20,8 @@ use crate::vtable::VTable;
 
 #[derive(Clone, Copy, Debug, PartialEq, Eq)]
 pub struct Table<'a> {
-    buf: &'a [u8],
-    loc: usize,
+    pub buf: &'a [u8],
+    pub loc: usize,
 }
 
 impl<'a> Table<'a> {
